### PR TITLE
Rework logging system

### PR DESCRIPTION
**Pull request by nto**
- defer printf formatting until the message is actually needed,
   otherwise some cpu time is wasted on verbose logs that are never
   displayed

 - properly handle multi-line log messages

 - replace macros with inline functions, making them accessible from
   luajit bindings

 - add info log level

### CODE REVIEW

File: include/upipe/uprobe.h [hunk 1]
```diff
@@ -300,20 +300,24 @@ static inline int uprobe_throw_next(struct uprobe *uprobe, struct upipe *upipe,
...
-static inline void uprobe_log(struct uprobe *uprobe, struct upipe *upipe,
-                              enum uprobe_log_level level, const char *msg)
+static inline void uprobe_vlog(struct uprobe *uprobe, struct upipe *upipe,
+                               enum uprobe_log_level level,
+                               const char *format,
+                               va_list args)
 {
     struct ulog ulog;
-    ulog_init(&ulog, level, msg);
+    ulog_init(&ulog, level, format);
```
Comment (quarium) [inline include/upipe/uprobe.h:317]: Maybe you should remove `ulog_init` as it doesn't initialise `ulog.args`?

File: include/upipe/ulog.h [hunk 2]
```diff
@@ -78,13 +82,33 @@ struct ulog {
  */
```
Comment (quarium) [inline include/upipe/ulog.h:82]: `msg` need to be updated

File: include/upipe/ulog.h [hunk 3]
```diff
@@ -78,13 +82,33 @@ struct ulog {
...
+{
+    va_list args;
+
+    va_copy(args, ulog->args);
+    int ret = vsnprintf(buf, size, ulog->format, args);
+    va_end(args);
+    return ret;
+}
+
+UBASE_PRAGMA_CLANG(diagnostic pop)
```
Comment (quarium) [inline include/upipe/ulog.h:105]: It would be nice to have a comment explaining why this is needed

File: include/upipe/ulog.h [hunk 4]
```diff
@@ -78,13 +82,33 @@ struct ulog {
...
+
+static inline int ulog_msg_len(struct ulog *ulog)
```
Comment (quarium) [inline include/upipe/ulog.h:107]: Can you add the documentation?

File: include/upipe/ulog.h [hunk 5]
```diff
@@ -78,13 +82,33 @@ struct ulog {
...
     ulog->level = level;
-    ulog->msg = msg;
+    ulog->format = format;
     ulist_init(&ulog->prefixes);
 }
 
+UBASE_PRAGMA_CLANG(diagnostic push)
+UBASE_PRAGMA_CLANG(diagnostic ignored "-Wformat-nonliteral")
+
+static inline int ulog_msg_print(struct ulog *ulog, char *buf, size_t size)
```
Comment (quarium) [inline include/upipe/ulog.h:95]: Can you add the documentation?

File: include/upipe/ulog.h (hunk 4 above)
Comment (nto) [reply include/upipe/ulog.h:123]: Done.

File: include/upipe/ulog.h (hunk 5 above)
Comment (nto) [reply include/upipe/ulog.h:106]: Done.

File: include/upipe/ulog.h (hunk 3 above)
Comment (nto) [reply include/upipe/ulog.h]: Done.

File: include/upipe/ulog.h (hunk 2 above)
Comment (nto) [reply include/upipe/ulog.h:83]: Done.

File: include/upipe/uprobe.h (hunk 1 above)
Comment (nto) [reply include/upipe/uprobe.h]: `ulog.args` is initialized now.

**Pull request merged**